File: Haskell/Task5.fs
data Tree a = Empty | Node a Integer (Tree a) (Tree a) deriving Show

insert Empty x = Node x 1 Empty Empty
insert (Node a h l r) x
	| a > x = Node a h (insert l x) r
	| a < x = Node a h l (insert r x)
	| otherwise = Node a h l r


getH Empty = 0
getH (Node _ h _ _) = h

setH Empty = Empty
setH (Node a _ l r) = Node a (max (getH l) (getH r) +1) l r

LR (Node a h (Node al hl ll lr) r) = setH (Node al -1 ll (setH (Node a -1 lr r))) 
RR (Node a h l (Node ar hr rl rr)) = setH (Node ar -1 (setH (Node a -1 l rl)) rr)

--balance (Node a h l@(Node al hl ll rl) r@(Node ar hr lr rr)) 
--	| hr+1<hl = Node al hl (balance ll) (balance (Node a hl-1 rl r))
--	| hl+1<hr = Node ar hr (balance (Node a hr-1 l lr)) (balance rr)
--	| otherwise = Node a h (balance l) (balance r)
--balance (Node a h (Node al hl ll rl) Empty) = Node al h (balance ll) (balance (Node a hl rl Empty))
--balance (Node a h Empty (Node ar hr lr rr)) = Node ar h (balance (Node a hr Empty lr)) (balance rr)
--balance Empty = Empty

File: Haskell/Task5.hs
data Tree a = Empty | Node a Integer (Tree a) (Tree a) deriving Show

insert Empty x = Node x 1 Empty Empty
insert nd@(Node a h l r) x
	| a > x = setH (Node a h (insert l x) r)
	| a < x = setH (Node a h l (insert r x))
	| otherwise = nd

setH (Node a _ l r) = Node a (max (getH l) (getH r) +1) l r
setH Empty = Empty

getH (Node _ h _ _) = h
getH Empty = 0

-- 4 type of rotate
balance nd@(Node a h l r)
	| (dbh nd)-1 > 0 && dbh r >  0 = lR nd
	| (dbh nd)-1 > 0 && dbh r <= 0 = lR (Node a h l (rR r))
	| (dbh nd)+1 < 0 && dbh l <  0 = rR nd
	| (dbh nd)+1 < 0 && dbh l >= 0 = rR (Node a h (lR l) r)
	| otherwise = nd

  where
		

	lR (Node a h (Node al hl ll lr) r) = setH (Node al 0 ll (setH (Node a 0 lr r)))
	lR Empty = Empty 
	rR (Node a h l (Node ar hr rl rr)) = setH (Node ar 0 (setH (Node a 0 l rl)) rr)
	rR Empty = Empty

	dbh (Node _ _ l r) = (getH l) - (getH r)	
	dbh Empty = 0
	
